docs/proposals/PROPOSAL_D.md

```diff
@@ -0,0 +1,480 @@
+      },
+      "hardware.gpu.driver.version": {
+        "label": "gpu driver version"
+      },
+      "hardware.gpu.cuda": {
+        "label": "gpu cuda"
+      },
+      "hardware.gpu.cuda.version": {
+        "label": "cuda version"
+      },
+      "communication": {
+        "label": "communication"
+      },
+      "communication.framework": {
+        "label": "communication framework"
+      }
+    },
+    "edges": [
+      {
+        "source": "mpi",
+        "target": "mpi.version",
+        "relation": "contains"
+      },
+      {
+        "source": "mpi",
+        "target": "mpi.implementation",
+        "relation": "contains"
+      },
+      {
+        "source": "mpi",
+        "target": "mpi.portability",
+        "relation": "contains"
+      },
+      {
+        "source": "mpi.portability",
+        "target": "mpi.portability.optimization",
+        "relation": "contains"
+      },
+      {
+        "source": "mpi.portability",
+        "target": "mpi.portability.mode",
+        "relation": "contains"
+      },
+      {
+        "source": "os",
+        "target": "os.name",
+        "relation": "contains"
+      },
+      {
+        "source": "os",
+        "target": "os.release",
+        "relation": "contains"
+      },
+      {
+        "source": "os",
+        "target": "os.vendor",
+        "relation": "contains"
+      },
+      {
+        "source": "os",
+        "target": "os.version",
+        "relation": "contains"
+      },
+      {
+        "source": "hardware",
+        "target": "hardware.gpu",
+        "relation": "contains"
+      },
+      {
+        "source": "hardware.gpu",
+        "target": "hardware.gpu.enabled",
+        "relation": "contains"
+      },
+      {
+        "source": "hardware.gpu",
+        "target": "hardware.gpu.driver",
+        "relation": "contains"
+      },
+      {
+        "source": "hardware.gpu.driver",
+        "target": "hardware.gpu.driver.version",
+        "relation": "contains"
+      },
+      {
+        "source": "hardware.gpu",
+        "target": "hardware.gpu.architecture",
+        "relation": "contains"
+      },
+      {
+        "source": "hardware.gpu",
+        "target": "hardware.gpu.version",
+        "relation": "contains"
+      },
+      {
+        "source": "hardware.gpu",
+        "target": "hardware.gpu.cuda",
+        "relation": "contains"
+      },
+      {
+        "source": "hardware.gpu.cuda",
+        "target": "hardware.gpu.cuda.version",
+        "relation": "contains"
+      },
+      {
+        "source": "communication",
+        "target": "communication.framework",
+        "relation": "contains"
+      }
+    ],
+    "metadata": {
+      "version": "0.0.0",
+      "source": "https://github.com/supercontainers/compspec"
+    }
+  }
+}
+```
+
+## Compatibility Artifact
+
+The compatibility artifact would include a subset of metadata attributes that are known to a schema, along with versions and links to the schemas used.
```

Comment:
I am not sure if it's safe. It's realted to the previous comment about the schema. If we fail to download the schema we are not able to verify the fields. I think the binary from our tool should already include all schemas to avoid potential break at runtime.

Reply:
Tools are not required to use or parse schemas. The alternative to hard code a graph, and a large one, and in _every_ single artifact that needs to be changed (and it may not even be used). That would be hugely challenging to update, and also wasteful in size if the graph is ultimately not needed.

Reply:
I am not saying to hard code a graph in the artifact. I am only saying that if we agree to release an official tool under OCI (with whatever scope) it would be good that the tool already knows schemas (at build time) from all organizations it incorporates.

> Tools are not required to use or parse schemas.

True, but they can.

It is just food for thought. I am not asking to change this in the proposal.
Once we have discussions about the tooling, its scope, plugins, releases etc. it should clarify itself.

Reply:
That's an interesting idea that it might know them beforehand - how would that work when there could be updated versions, etc.? What is the difference / time saved for having a schema built in vs. loaded when needed?

Reply:
> how would that work when there could be updated versions, etc.?

Depends on the architecture and the release process we agree on. Most likely all the updates of schema would be included in the next release of the tool. Like in k8s old in-tree CCM approach.

> What is the difference / time saved for having a schema built in vs. loaded when needed?

I don't think it's gonna be a big difference from the time/performance point of view. It's mostly for resilience.